google-cloud-storage/src/test/java/com/google/cloud/storage/StorageImplMockitoTest.java

```diff
@@ -1637,4 +1661,2364 @@ public void testWriterFailure() {
+    storage =
+        options
+            .toBuilder()
+            .setRetrySettings(ServiceOptions.getDefaultRetrySettings())
+            .build()
+            .getService();
+    initializeServiceDependentObjects();
+    try {
+      storage.get(blob);
+      Assert.fail();
+    } catch (StorageException ex) {
+      Assert.assertNotNull(ex.getMessage());
+    }
+  }
+
+  @Test
+  public void testRuntimeException() {
+    BlobId blob = BlobId.of(BUCKET_NAME1, BLOB_NAME1);
+    String exceptionMessage = "Artificial runtime exception";
+    doThrow(new RuntimeException(exceptionMessage))
+        .doThrow(UNEXPECTED_CALL_EXCEPTION)
+        .when(storageRpcMock)
+        .get(blob.toPb(), EMPTY_RPC_OPTIONS);
+
+    storage =
+        options
+            .toBuilder()
+            .setRetrySettings(ServiceOptions.getDefaultRetrySettings())
+            .build()
+            .getService();
+    try {
+      storage.get(blob);
+      Assert.fail();
+    } catch (StorageException ex) {
+      Assert.assertNotNull(ex.getMessage());
+    }
+  }
+
+  @Test
+  public void testWriterWithSignedURL() throws MalformedURLException {
+    doReturn("upload-id").doThrow(UNEXPECTED_CALL_EXCEPTION).when(storageRpcMock).open(SIGNED_URL);
+    initializeService();
+    WriteChannel writer = new BlobWriteChannel(options, new URL(SIGNED_URL));
+    assertNotNull(writer);
+    assertTrue(writer.isOpen());
+  }
+
+  @Test
+  public void testV4PostPolicy() {
+    ServiceAccountCredentials credentials =
+        ServiceAccountCredentials.newBuilder()
+            .setClientEmail(ACCOUNT)
+            .setPrivateKey(privateKey)
+            .build();
+    storage = options.toBuilder().setCredentials(credentials).build().getService();
+
+    PostPolicyV4.PostFieldsV4 fields =
+        PostPolicyV4.PostFieldsV4.newBuilder().setAcl("public-read").build();
+    PostPolicyV4.PostConditionsV4 conditions =
+        PostPolicyV4.PostConditionsV4.newBuilder()
+            .addContentTypeCondition(PostPolicyV4.ConditionV4Type.MATCHES, "image/jpeg")
+            .build();
+
+    // test fields and conditions
+    PostPolicyV4 policy =
+        storage.generateSignedPostPolicyV4(
+            BlobInfo.newBuilder("my-bucket", "my-object").build(),
+            7,
+            TimeUnit.DAYS,
+            fields,
+            conditions);
+
+    Map<String, String> outputFields = policy.getFields();
+
+    assertTrue(outputFields.containsKey("x-goog-date"));
+    assertTrue(outputFields.containsKey("x-goog-credential"));
+    assertTrue(outputFields.containsKey("x-goog-signature"));
+    assertEquals(outputFields.get("x-goog-algorithm"), "GOOG4-RSA-SHA256");
+    assertEquals(outputFields.get("content-type"), "image/jpeg");
+    assertEquals(outputFields.get("acl"), "public-read");
+    assertEquals(outputFields.get("key"), "my-object");
+    assertEquals("https://storage.googleapis.com/my-bucket/", policy.getUrl());
+
+    // test fields, no conditions
+    policy =
```

Comment:
Reassigning local variables is a code smell that suggests this could be broken up into multiple tests.

Reply:
agree, done. 
Please take a look.

Reply:
@elharo, it would be awesome if you take another look at this change before the New Year